### PR TITLE
Scala marshalling DSL + XSD parser / validator

### PR DESCRIPTION
**Pull request by jypma**
Also throwing an XSD parser in here to test out how comfortable the scala API is.
- [x] Find out why Scala won't type infer the lambda argument types right now, where Java does.

### CODE REVIEW

File: ts-reaktive-marshal-scala/src/test/scala/com/tradeshift/reaktive/marshal/scaladsl/XMLProtocolsSpec.scala [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+class XMLProtocolsSpec extends FunSpec {
+  case class DTO1(l: Long, i: Option[Int], s: Seq[String])
+  
+  describe("A protocol mapping both attributes and child tags") {
+    import Protocols._
+    import XMLProtocols._
+    import StringMarshallable._
+    
+    val dto1proto =
```
Comment (jypma) [inline ts-reaktive-marshal-scala/src/test/scala/com/tradeshift/reaktive/marshal/scaladsl/XMLProtocolsSpec.scala]: @jrudolph @ktoso So this is how a case class would be mapped to XML. In the Java equivalent, the compiler will infer all the function types just fine. But Scala insists on having the types declared... any idea why?

Comment (jypma) [inline ts-reaktive-marshal-scala/src/test/scala/com/tradeshift/reaktive/marshal/scaladsl/XMLProtocolsSpec.scala]: got it. Forgot a minor detail: there _is_ a specific type on the `dto1proto` itself on the java side. Added that, all good.

File: ts-reaktive-marshal-scala/src/main/scala/com/tradeshift/reaktive/marshal/scaladsl/XMLProtocols.scala [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+  /**
+   * Writes a tag and one child element (tag or attribute), where the result of this tag is the result of the single child.
+   */
+  def tag[T](name: QName, p1: WriteProtocol[XMLEvent, T]): TagWriteProtocol[T] =
+    XMLProtocol.tag(name, p1)
+  
+  /**
+   * Reads and writes a tag and one child element (tag or attribute) using [p1], using [f] to create the result on reading, getting values using [g1] for writing.
+   */
+  def tag[T,F1](name: QName, p1: Protocol[XMLEvent, F1], f: F1 => T, g1: T => F1): TagProtocol[T] =
```
Comment (ktoso) [inline ts-reaktive-marshal-scala/src/main/scala/com/tradeshift/reaktive/marshal/scaladsl/XMLProtocols.scala]: please better names for those arguments ;)
Scala people would use them as `protocol = ...`

File: ts-reaktive-marshal-scala/src/main/scala/com/tradeshift/reaktive/marshal/scaladsl/XMLProtocols.scala [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+import com.tradeshift.reaktive.xml.TagReadProtocol
+import com.tradeshift.reaktive.xml.TagWriteProtocol
+import com.tradeshift.reaktive.marshal.WriteProtocol
+import javax.xml.stream.events.Namespace
+import com.tradeshift.reaktive.xml.AttributeProtocol
+import com.tradeshift.reaktive.xml.BodyProtocol
+import javax.xml.stream.events.Characters
+import com.tradeshift.reaktive.xml.AnyAttributeProtocol
+
+object XMLProtocols {
```
Comment (ktoso) [inline ts-reaktive-marshal-scala/src/main/scala/com/tradeshift/reaktive/marshal/scaladsl/XMLProtocols.scala]: Protocol I think

File: ts-reaktive-marshal-scala/src/test/scala/com/tradeshift/reaktive/marshal/scaladsl/XMLProtocolsSpec.scala [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+            ),
+            vector(
+                tag(qname("s"),
+                    body
+                )
+            ),
+            DTO1,
+            _.l,
+            _.i,
+            _.s
```
Comment (ktoso) [inline ts-reaktive-marshal-scala/src/test/scala/com/tradeshift/reaktive/marshal/scaladsl/XMLProtocolsSpec.scala]: hmmm this is actually pretty difficult to read, so tag is restricted to only have 3 fields and you'll have to generate the rest of them now?

Comment (ktoso) [inline ts-reaktive-marshal-scala/src/test/scala/com/tradeshift/reaktive/marshal/scaladsl/XMLProtocolsSpec.scala]: For reference a more advanced API would likely use http://julien-truffaut.github.io/Monocle/optics/lens.html but I guess that's out of scope for now, dont worry about it too much

Comment (jypma) [inline ts-reaktive-marshal-scala/src/test/scala/com/tradeshift/reaktive/marshal/scaladsl/XMLProtocolsSpec.scala]: `Lens` is certainly interesting, but I'd have two questions:
1. For a class with say 10 fields, isn't invoking `copy` 10 times going to be noticably slower than invoking a single constructor?
2. Weren't macros going away in Dotty / Scala? I think I heared Martin O. say sth like that.

File: ts-reaktive-marshal-scala/src/main/scala/com/tradeshift/reaktive/marshal/scaladsl/XMLProtocols.scala (hunk 2 above)
Comment (jypma) [reply ts-reaktive-marshal-scala/src/main/scala/com/tradeshift/reaktive/marshal/scaladsl/XMLProtocols.scala]: certainly. Once we generate them it should probably be `protocol1`, `apply`, `getter1`, etc.

**Pull request merged**